src/vehicle_autodrive.cpp

```diff
@@ -1,25 +1,1202 @@
+ * node, current speed and amount of steering we can predict where the vehicle's pivot point
+ * will end up in 1 turn and connect the 2 nodes.
+ *
+ * In order to keep the search space small we only consider paths traveled at constant speed
+ * (otherwise we'd have to add speed as an extra attribute in every node). When we need to
+ * compute a new path we first try to use the maximum safe autodrive speed (which is typically
+ * much less than the vehicle's max speed). If we can't find a path we cut the speed in half and
+ * try again, repeating until we succeed or reach the minimum speed. When the target speed is
+ * different from the current speed of the vehicle we try to estimate how long it will take to
+ * reach that speed and take that into account when simulating the first few edges (they'll be
+ * shorter or longer than the rest).
+ *
+ * Once we have generated a sequence of nodes to follow, we do not invoke A* again until we reach
+ * the next OMT. We simply cache the path and on every subsequent call to do_autodrive() we
+ * check where we are on that path, what node we're supposed to reach next and what action we
+ * should use to get there. However, sometimes the simplified simulation used to compute the
+ * search graph fails to accurately predict where the vehicle will end up. This could be because
+ * there was too much error in an approximation (such as time to accelerate) or the driving
+ * conditions changed (ran over a mound of dirt that interfered with steering). When such a
+ * discrepancy is detected, autodrive will fall back to "safe mode", cutting the speed to the
+ * minimum (1 tile per second) and recomputing the path.
+ *
+ * Since the navigation graph is cached and not updated in response to dynamic obstacles (such
+ * as animals) autodrive will also perform collision detection at every turn, before taking
+ * any action that may end the turn. If a possible collision is detected autodrive will enter
+ * "safe mode", reducing speed and recomputing. If the speed is already at minimum it will
+ * abort instead with a warning message.
+ *
+ * In most cases when autodrive aborts it will not automatically stop the vehicle. Stopping
+ * may be undesirable since nearby enemies could get into the vehicle. It is up to the player
+ * to either stop or continue driving manually for a while. However, when aborting due to a
+ * collision (which happened despite collision detection, i.e. there was a simulation discrepancy
+ * near an obstacle) the brakes will be engaged.
+ *
+ * In order to prevent exploiting the system to navigate in the dark, autodrive will also
+ * perform visibility checks. Not being able to see anything in front of the vehicle will
+ * immediately cancel (or fail to start) autodrive. If the driver has only partial visibility
+ * of the front of the vehicle, safe mode will engage limiting the speed. Furthermore, the
+ * driver needs to either see or have map memory for the map tiles in the current OMT; all
```

Comment:
How does this work if it's an NPC driver?

Reply:
NPCs don't have map memory so they need full visibility of the current OMT (meaning they can only drive during the day). If we really want to make NPC autodriving a reality, this probably needs to be changed (either they share the player's memory, or they get their own memory, or we drop this requirement for them).

Reply:
I can see sharing the player's memory _if_ the player is in the vehicle.

Reply:
Right now, the player would have to be in the vehicle anyway. The map::ter(), map::furn(), map::veh_at() and game::critter_at() calls in check_drivable() don't work outside of the reality bubble. Also NPC and vehicle objects are unloaded when exiting the bubble.
In order to have NPCs driving on their own, they'd either need their own mini reality bubble (probably with partial simulation due to performance concerns) or have their travel completely abstracted (they way it is now when they travel on foot).